src/segment/mod.rs

```diff
@@ -296,11 +319,169 @@ impl Segment {
+                    .contains_prefix(key, prefix_extractor.as_ref())
+                    .unwrap_or(true);
+            }
+        }
+
+        // Try loading filter block from disk
+        if let Some(filter_block_handle) = &self.regions.filter {
+            if let Ok(block) = self.load_block(
+                filter_block_handle,
+                BlockType::Filter,
+                CompressionType::None,
+            ) {
+                if let Ok(filter) = StandardBloomFilterReader::new(&block.data) {
+                    #[cfg(feature = "metrics")]
+                    self.metrics.bloom_filter_queries.fetch_add(1, Relaxed);
+
+                    return !filter
+                        .contains_prefix(key, prefix_extractor.as_ref())
+                        .unwrap_or(true);
+                }
+            }
+        }
+
+        false
+    }
+
+    /// Extracts the common prefix from a range's start and end bounds for bloom filter checking.
+    /// Assumes a prefix extractor is configured.
+    fn extract_common_prefix_for_bloom_filter<R: RangeBounds<UserKey>>(
+        &self,
+        range: &R,
+        prefix_extractor: &dyn PrefixExtractor,
+    ) -> Option<Vec<u8>> {
+        let start_key = match range.start_bound() {
+            Bound::Included(key) | Bound::Excluded(key) => Some(key.as_ref()),
+            Bound::Unbounded => None,
+        };
+
+        let end_key = match range.end_bound() {
+            Bound::Included(key) | Bound::Excluded(key) => Some(key.as_ref()),
+            Bound::Unbounded => None,
+        };
+
+        match (start_key, end_key) {
+            (Some(start), Some(end)) => {
+                // Extract prefixes using the prefix extractor
+                let start_prefixes: Vec<_> = prefix_extractor.extract(start).collect();
+                let end_prefixes: Vec<_> = prefix_extractor.extract(end).collect();
+
+                // If either key is out of domain (no prefixes), we can't optimize
+                if start_prefixes.is_empty() || end_prefixes.is_empty() {
+                    return None;
+                }
+
+                // Check if the first prefix of each is the same
+                // For most extractors, there's only one prefix
+                if let (Some(start_prefix), Some(end_prefix)) =
+                    (start_prefixes.first(), end_prefixes.first())
+                {
+                    if start_prefix == end_prefix {
+                        Some(start_prefix.to_vec())
+                    } else {
+                        // Different prefixes, no common prefix
+                        None
+                    }
+                } else {
+                    None
+                }
+            }
+            (Some(key), None) | (None, Some(key)) => {
+                // If only one bound exists, use its extracted prefix
+                let prefixes: Vec<_> = prefix_extractor.extract(key).collect();
+                prefixes.first().map(|p| p.to_vec())
```

Comment:
There is a lot of collect() and first() here and there... why not just next()? So we avoid the vector allocation.